plugins/com.google.cloud.tools.eclipse.appengine.deploy/src/com/google/cloud/tools/eclipse/appengine/deploy/WarPublisher.java

```diff
@@ -85,28 +79,46 @@ private static void publish(IProject project, IPath destination, boolean explode
     }
   }
 
-  private static List<IModuleResource> computeWebFragmentJarResources(
-      J2EEFlexProjDeployable deployable, IProgressMonitor monitor) throws CoreException {
-    List<IModuleResource> webFragmentJars = new ArrayList<>();
-
-    IModule[] childModules = deployable.getChildModules();
-    for (IModule module : childModules) {
-      if ("jst.webfragment".equals(module.getModuleType().getId())
-          && module.getName().toLowerCase(Locale.US).endsWith(".jar")) {
-        ProjectModule projectModule = (ProjectModule)
-            module.loadAdapter(ProjectModule.class, monitor);
-
-        for (IModuleResource resource : projectModule.members()) {
-          IFile file = resource.getAdapter(IFile.class);
-          File file2 = resource.getAdapter(File.class);
-          if (file != null) {
-            webFragmentJars.add(new ModuleFile(file, resource.getName(), new Path("WEB-INF/lib")));
-          } else if (file2 != null) {
-            webFragmentJars.add(new ModuleFile(file2, resource.getName(), new Path("WEB-INF/lib")));
-          }
+  private static IModuleResource[] flattenResources(
+      IProject project, IProgressMonitor monitor) throws CoreException {
+    List<IModuleResource> resources = new ArrayList<>();
+
+    IModule[] modules = ServerUtil.getModules(project);
+    for (IModule module : modules) {
+      ModuleDelegate delegate = (ModuleDelegate) module.loadAdapter(ModuleDelegate.class, monitor);
+      if (delegate == null) {
+        continue;
+      }
+
+      // module references can either be as members or child modules (http://eclip.se/467759)
+      Collections.addAll(resources, delegate.members());
+
+      // now handle web fragment child modules, if they exist
+      for (IModule child : delegate.getChildModules()) {
+        ModuleDelegate childDelegate = (ModuleDelegate)
+            child.loadAdapter(ModuleDelegate.class, monitor);
+        IWebFragmentModule webFragmentModule = (IWebFragmentModule)
+            child.loadAdapter(IWebFragmentModule.class, monitor);
+        if (childDelegate == null || webFragmentModule == null || !webFragmentModule.isBinary()) {
+          continue;
+        }
+
+        // e.g., "WEB-INF/lib/spring-web-4.3.6.RELEASE.jar"
+        IPath zip = new Path(delegate.getPath(child));
```

Comment:
It would be helpful to indicate that this is a destination, and not the actual zip.